torchtune/datasets/_packed.py

```diff
@@ -119,156 +121,170 @@ def _pack(self) -> None:
+        # Pad tokens
+        padded_tokens = F.pad(
+            pack["tokens"],
+            (0, self.max_seq_len - len(pack["tokens"])),
+            value=padding_idx,
         )
-        # Do not go beyond max_seq_len - 1
-        input_pos_pad = input_pos_pad.clamp(max=self.max_seq_len - 1)
-        input_pos = torch.cat(
-            [
-                torch.tensor(input_pos),
-                input_pos_pad,
-            ]
+
+        # Pad labels
+        padded_labels = F.pad(
+            pack["labels"],
+            (0, self.max_seq_len - len(pack["labels"])),
+            value=CROSS_ENTROPY_IGNORE_IDX,
         )
 
-        assert tokens.shape == labels.shape == input_pos.shape
-        assert tokens.shape[0] == mask.shape[0]
+        # Pad input_pos continuing the sequence from last value
+        # in input_pos
+        # e.g. [0 1 2] -> [0 1 2 3 4 5] for self.max_seq_len = 6
+        num_range = torch.arange(
+            pack["input_pos"][-1] + 1,
+            pack["input_pos"][-1] + self.max_seq_len - len(pack["input_pos"]) + 1,
+        )
+        # Clamp to max_seq_len - 1 to avoid out of bounds error
+        clamped_num_range = torch.clamp(num_range, 0, self.max_seq_len - 1)
+        padded_input_pos = torch.cat([pack["input_pos"], clamped_num_range])
 
         return {
-            "tokens": tokens,
-            "labels": labels,
-            "mask": mask,
-            "input_pos": input_pos,
+            "tokens": padded_tokens,
+            "labels": padded_labels,
+            "input_pos": padded_input_pos,
+            "seq_lens": pack["seq_lens"],  # seq_len is untouched
         }
 
-    def __len__(self):
+    def __len__(self) -> int:
         return len(self.packs)
 
-    def __getitem__(self, index: int) -> Dict[str, List[int]]:
-        return self.packs[index]
+    def __getitem__(self, idx: int) -> Dict[str, torch.Tensor]:
+        """Constructs the attention mask on-the-fly and returns whole sample."""
+        current_pack = self.packs[idx]
+
+        num_samples_in_pack = len(current_pack["seq_lens"])
+        total_seq_len = 0
+
+        block_attn_masks = []
+
+        for i, seq_len in enumerate(current_pack["seq_lens"]):
+            total_seq_len += seq_len
+
+            # Append lower triangular matrix for causal mask
+            block_attn_masks.append(
+                torch.tril(torch.ones(seq_len, seq_len, dtype=torch.bool))
+            )
+
+            # If we're at the last sample and the total seq len is less than the max seq len,
+            # we need to pad with identity matrix for the remainder
+            if i == num_samples_in_pack - 1 and total_seq_len < self.max_seq_len:
+                block_attn_masks.append(
+                    torch.eye(
+                        self.max_seq_len - total_seq_len,
+                        self.max_seq_len - total_seq_len,
+                        dtype=torch.bool,
+                    )
+                )
```

Comment:
Not a major blocker for this PR, but just fyi it is probably faster to do this in a list comprehension instead of append. See e.g. [here](https://gist.github.com/ebsmothers/707ee5a260b0e25b820a19b2e634e921). If number of samples per pack is generally small prob not a huge deal, but we will be doing this for every sample so it may add up

Reply:
Filed #1130 for a follow-up